File: src/formlabs_local_mcp/paths.py
"""Guard rails for every tool that hands a file path to PreFormServer.

PreFormServer reads and writes whatever absolute path it is given, as the
user running it. A prompt-injected model could otherwise be talked into
writing `~/.ssh/authorized_keys` or reading `/etc/passwd`. Three checks keep
that surface small:

1. The path must be absolute (PreFormServer rejects anything else anyway).
2. It must resolve (symlinks included) to somewhere under an allowed root.
   Default root: the user's home directory.
3. No path component may be hidden (start with ".") unless explicitly
   allowed. That keeps `~/.ssh`, `~/.aws`, `~/.config` and friends off limits
   even though they live under home.
4. The extension must match what the tool is for (`.stl` in, `.form` out).
"""

from __future__ import annotations

import os
from pathlib import Path

from formlabs_local_mcp.config import Config

MODEL_EXTENSIONS = frozenset({".stl", ".obj", ".3mf", ".step", ".stp", ".form"})
FORM_EXTENSIONS = frozenset({".form"})
FPS_EXTENSIONS = frozenset({".fps"})
IMAGE_EXTENSIONS = frozenset({".png", ".webp"})


class PathNotAllowed(ValueError):
    """Raised when a path fails one of the checks above."""


def _check(path: str, config: Config, extensions: frozenset[str], must_exist: bool) -> Path:
    if not path or not path.strip():
        raise PathNotAllowed("A file path is required.")
    p = Path(os.path.expanduser(path))
    if not p.is_absolute():
        raise PathNotAllowed(
            f"Path must be absolute (got {path!r}). Resolve it against the working "
            "directory first; PreFormServer does not accept relative paths."
        )
    resolved = p.resolve()
    if resolved.suffix.lower() not in extensions:
        allowed = ", ".join(sorted(extensions))
        raise PathNotAllowed(f"{path!r} must end in one of: {allowed}.")
    if not any(resolved == root or root in resolved.parents for root in config.allowed_paths):
        roots = ", ".join(str(r) for r in config.allowed_paths)
        raise PathNotAllowed(
            f"{path!r} is outside the allowed directories ({roots}). "
            "Set FORMLABS_ALLOWED_PATHS to add more."
        )
    if not config.allow_hidden_paths:
        rel_parts = resolved.parts
        for root in config.allowed_paths:
            if resolved == root or root in resolved.parents:
                rel_parts = resolved.relative_to(root).parts
                break
        hidden = [part for part in rel_parts if part.startswith(".")]
        if hidden:
            raise PathNotAllowed(
                f"{path!r} passes through a hidden directory or file ({hidden[0]}). "
                "Set FORMLABS_ALLOW_HIDDEN_PATHS=1 if that is intentional."
            )
    if must_exist and not resolved.is_file():
        raise PathNotAllowed(f"{path!r} does not exist or is not a file.")
    if not must_exist and not resolved.parent.is_dir():
        raise PathNotAllowed(f"The directory for {path!r} does not exist.")
    return resolved


def input_path(path: str, config: Config, extensions: frozenset[str] = MODEL_EXTENSIONS) -> str:
    """Validate a path PreFormServer will read. Returns the resolved string."""
    return str(_check(path, config, extensions, must_exist=True))


def output_path(path: str, config: Config, extensions: frozenset[str]) -> str:
    """Validate a path PreFormServer will write. Returns the resolved string."""
    return str(_check(path, config, extensions, must_exist=False))

File: src/paths.ts
/**
 * Guard rails for every tool that hands a file path to PreFormServer.
 *
 * PreFormServer reads and writes whatever absolute path it is given, as the
 * user running it. A prompt-injected model could otherwise be talked into
 * writing `~/.ssh/authorized_keys` or reading `/etc/passwd`. Four checks keep
 * that surface small:
 *
 * 1. The path must be absolute (PreFormServer rejects anything else anyway).
 * 2. It must resolve (symlinks included) to somewhere under an allowed root.
 *    Default root: the user's home directory.
 * 3. No path component under the root may be hidden (start with ".") unless
 *    explicitly allowed, which keeps ~/.ssh, ~/.aws and friends off limits.
 * 4. The extension must match what the tool is for (models in, .form out).
 */

import { existsSync, realpathSync, statSync } from "node:fs";
import path from "node:path";
import { expandHome, type Config } from "./config.js";

export const MODEL = [".stl", ".obj", ".3mf", ".step", ".stp", ".form"];
export const FORM = [".form"];
export const FPS = [".fps"];
export const IMAGE = [".png", ".webp"];

export class PathNotAllowed extends Error {}

function realpathOrSelf(p: string): string {
  try {
    return realpathSync.native(p);
  } catch {
    return p;
  }
}

/** Resolve as much of `p` as exists, so a not-yet-created output file still gets symlink-checked. */
function resolveDeep(p: string): string {
  const parent = path.dirname(p);
  if (existsSync(p)) return realpathOrSelf(p);
  if (parent === p) return p;
  return path.join(resolveDeep(parent), path.basename(p));
}

function isUnder(target: string, root: string): boolean {
  const rel = path.relative(root, target);
  return rel === "" || (!rel.startsWith("..") && !path.isAbsolute(rel));
}

function check(raw: string, cfg: Config, extensions: string[], mustExist: boolean): string {
  if (!raw || !raw.trim()) throw new PathNotAllowed("A file path is required.");
  const expanded = expandHome(raw, cfg.home);
  if (!path.isAbsolute(expanded)) {
    throw new PathNotAllowed(
      `Path must be absolute (got ${JSON.stringify(raw)}). Resolve it against the working directory first; PreFormServer does not accept relative paths.`,
    );
  }
  const resolved = resolveDeep(path.normalize(expanded));
  const ext = path.extname(resolved).toLowerCase();
  if (!extensions.includes(ext)) {
    throw new PathNotAllowed(`${JSON.stringify(raw)} must end in one of: ${extensions.join(", ")}.`);
  }
  const roots = cfg.allowedPaths.map(realpathOrSelf);
  const root = roots.find((r) => isUnder(resolved, r));
  if (root === undefined) {
    throw new PathNotAllowed(
      `${JSON.stringify(raw)} is outside the allowed directories (${roots.join(", ")}). Set FORMLABS_ALLOWED_PATHS to add more.`,
    );
  }
  if (!cfg.allowHiddenPaths) {
    const hidden = path.relative(root, resolved).split(path.sep).find((part) => part.startsWith("."));
    if (hidden) {
      throw new PathNotAllowed(
        `${JSON.stringify(raw)} passes through a hidden directory or file (${hidden}). Set FORMLABS_ALLOW_HIDDEN_PATHS=1 if that is intentional.`,
      );
    }
  }
  if (mustExist) {
    if (!existsSync(resolved) || !statSync(resolved).isFile()) {
      throw new PathNotAllowed(`${JSON.stringify(raw)} does not exist or is not a file.`);
    }
  } else if (!existsSync(path.dirname(resolved)) || !statSync(path.dirname(resolved)).isDirectory()) {
    throw new PathNotAllowed(`The directory for ${JSON.stringify(raw)} does not exist.`);
  }
  return resolved;
}

/** Validate a path PreFormServer will read. Returns the resolved absolute path. */
export function inputPath(raw: string, cfg: Config, extensions: string[] = MODEL): string {
  return check(raw, cfg, extensions, true);
}

/** Validate a path PreFormServer will write. Returns the resolved absolute path. */
export function outputPath(raw: string, cfg: Config, extensions: string[]): string {
  return check(raw, cfg, extensions, false);
}
